### PR TITLE
58944 added requirements parsing fix

### LINKED ISSUE #58944: [BUG] pip.install requirements file parsing breaks on some files when used with `user:`
**Description**
When `pip.install` is called with a requirements file that has a comment that looks like a requirements link (i.e. `# via -r requirements.in, something`), and a `user` param, salt will parse the requirements file incorrectly and read the linked file as `requirementsin.in, something`.

**Setup**
To see the issue in latest salt:
 - setup a virtualenv for salt to use. (I used `/venv` in a container as the example)
 - setup a requirements file with contents: `requests  # via -r requirements.in, foo`
 - call `pip.install` with the venv, requirements file, and a local user.

**Steps to Reproduce the behavior**
Full docker example:
```
docker run -i --rm amazonlinux:2 bash - <<BASH
set -eux
cd /root
yum install -y python3-devel
python3 -m pip install virtualenv salt==3002.1
python3 -m virtualenv -p python3 /venv
echo 'requests  # via -r requirements.in, foobar' > /root/requirements.txt

salt-call --local --skip-grains pip.install requirements=/root/requirements.txt user=nobody bin_env=/venv
BASH
```
Should see a stacktrace ending with: `FileNotFoundError: [Errno 2] No such file or directory: 'requirements.in, foobar'`

**Expected behavior**
expected `pip.install` to ignore the comments in requirements.txt

**Screenshots**
N/A

**Versions Report**
<details><summary>salt --versions-report</summary>

```
Salt Version:                                                                                                      
           Salt: 3002.1                                                                                            
                                                         
Dependency Versions:                                                                                               
           cffi: Not Installed                  
       cherrypy: Not Installed                                                                                     
       dateutil: Not Installed                
      docker-py: Not Installed                                                                                     
          gitdb: Not Installed          
      gitpython: Not Installed                                                                                     
         Jinja2: 2.11.2                                                                                            
        libgit2: Not Installed                   
       M2Crypto: Not Installed                                                                                     
           Mako: Not Installed                      
   msgpack-pure: Not Installed                                                                                     
 msgpack-python: 1.0.0                                   
   mysql-python: Not Installed                           
      pycparser: Not Installed                                                                                     
       pycrypto: Not Installed                                                                                     
   pycryptodome: 3.9.9                                   
         pygit2: Not Installed                                                                                     
         Python: 3.7.9 (default, Aug 27 2020, 21:59:41)                                                            
   python-gnupg: Not Installed                                                                                     
         PyYAML: 5.3.1                                   
          PyZMQ: 20.0.0  
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.3
 
System Versions:
           dist: amzn 2 
         locale: utf-8
        machine: x86_64
        release: 5.4.0-52-generic
         system: Linux
        version: Amazon Linux 2
```
</details>

**Additional context**
Wrote a fix and a test: https://github.com/saltstack/salt/pull/58945

### PR DESCRIPTION
**Pull request by jweede**
### What issues does this PR fix or reference?
Fixes: #58944 

### Previous Behavior
invalid requirements link parsing

### New Behavior
extracts requirements file links correctly.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

### CODE REVIEW

**Comment by waynew** [comment]
Hey @jweede, thanks for the PR! This definitely looks like an issue, and certainly one we should fix.

That being said, I'm not 100% sure if this fix is the one we really want :thinking:

@s0undt3ch -- do you have any opinion/thoughts here?

**Comment by s0undt3ch** [comment]
Hmm, we should either fix the regex or `.rsplit("#", 1)` each line. Or use pip's parsing machinery to parse the lines....
Shlex might not be the best approach....

We do have to look out for requirements that look like `git+https://domain.tld/git#egg=PackageName`....

**Comment by jweede** [comment]
Thanks @s0undt3ch. That makes sense. I assumed since pip docs and code mention `shlex` that would be enough. 

I read over their reqs parsing code a little closer this time and noticed they strip comments out via `COMMENT_RE = re.compile(r'(^|\s+)#.*$')`. I could rewrite the PR to use this and a version of the prior regex to find requirements links while ignoring comments.

**Comment by s0undt3ch** [comment]
@jweede that would be awesome!

File: tests/unit/modules/test_pip.py [hunk 1]
```diff
@@ -1365,12 +1365,17 @@ def __init__(self, filename):
...
-                        b"--requirement=requirements-2.txt\t# --requirements-2.txt\n\n"
+                        "four\n"
+                        "five\n"
+                        "--requirement=requirements-2.txt\t# --requirements-2.txt\n\n"
                     ),
                     "requirements-2.txt": b"""six""",
-                    "requirements-3.txt": b"""seven""",
+                    "requirements-3.txt": (
+                        b"# some comment\n"
+                        b"-e git+ssh://git.example.com/MyProject#egg=MyProject # the project\n"
```
Comment (s0undt3ch) [inline tests/unit/modules/test_pip.py:1375]: Let's duplicate this one, this time without the comment in front. Please  :)

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Review by s0undt3ch — DISMISSED** [review]
Superb!

**Review by cmcmarrow — DISMISSED** [review]

**Comment by s0undt3ch** [comment]
re-run all

**Comment by sagetherage** [comment]
@jweede can you rebase this against the `master` branch, please? we are no longer accepting changes on `develop` thank you!

**Comment by jweede** [comment]
@sagetherage rebased to master :+1:

**Review by Ch3LL — CHANGES_REQUESTED** [review]
looks like there is a merge conflict and this will also require a changelog entry.

**Comment by jweede** [comment]
@Ch3LL resolved the conflict and added a changelog entry as requested :+1:

**Review by Ch3LL — APPROVED** [review]

**Comment by sagetherage** [comment]
re-run pr-centosstream-8-x86_64-py3-pytest

**Review by s0undt3ch — APPROVED** [review]

**Pull request merged**